examples/qualcomm/oss_scripts/qwen/qwen.py

```diff
@@ -0,0 +1,396 @@
+            if generated_tokens[-1] == tokenizer.eos_token_id:
+                break
+    logging.info(
+        f"Result of Qwen {stage} with static cache:\n {tokenizer.decode(generated_tokens, skip_special_tokens=True)} \n\n\n"
+    )
+
+
+def compile(args):  # noqa: C901
+    skip_node_id_set, skip_node_op_set = parse_skip_delegation_node(args)
+
+    # ensure the working directory exist.
+    os.makedirs(args.artifact, exist_ok=True)
+
+    if not args.compile_only and args.device is None:
+        raise RuntimeError(
+            "device serial is required if not compile only. "
+            "Please specify a device serial by -s/--device argument."
+        )
+
+    model_id = HUGGING_FACE_REPO_IDS[args.decoder_model]
+    config =AutoConfig.from_pretrained(model_id)
+
+    device = "cpu"
+    batch_size = 1
+    dtype = "float32"
+    cache_implementation = "static"
+    attn_implementation = "eager"
+
+    # Set configs
+    config.max_seq_len = args.max_seq_len
+    config.ar_len = 1  # kv mode
+    config.max_batch_size = batch_size
+    # config.num_hidden_layers = 1
+    config.enable_spinquant_r3 = True
+
+    # Some config has head_dim provided that is different from equation below(e.g., qwen3)
+    if not hasattr(config, "head_dim"):
+        config.head_dim = config.hidden_size // config.num_attention_heads
+    modeling_qwen2.QWEN2_ATTENTION_CLASSES['eager'] = QCQwen2Attention
+    modeling_qwen2.Qwen2RotaryEmbedding.forward = bypass_rotary_embedding
+    model = AutoModelForCausalLM.from_pretrained(
+        model_id,
+        device_map=device,
+        torch_dtype=dtype,
+        config=config,
+        attn_implementation=attn_implementation,
+        generation_config=GenerationConfig(
+            use_cache=True,
+            cache_implementation=cache_implementation,
+            max_length=args.max_seq_len,
+            cache_config={
+                "batch_size": batch_size,
+                "max_cache_len": args.max_seq_len,
+            },
+        ),
+    ).eval()
+    model_wrapper = TorchExportableModuleWithStaticCache(model)
+    model_wrapper = initialize_r3_hadamard(model_wrapper)
+    model_wrapper = convert_linear_to_conv2d(model_wrapper)
+    model_wrapper = replace_qwen2_rms_norm_with_native_rms_norm(model_wrapper)
+    
+    tokenizer = AutoTokenizer.from_pretrained(model_id)
+
+    with torch.no_grad():
+        # Validate on model wrapper
+        calibrate(model_wrapper, "nn module", args.max_seq_len, tokenizer, args.prompt)
+
+    use_fp16 = True
+    fixed_point_type = {}
+    if args.ptq:
+        use_fp16 = False
+        if args.ptq == "8a8w":
+            fixed_point_type["io_type"] = torch.uint8
+            fixed_point_type["kv_type"] = torch.uint8
+        elif args.ptq in ("16a8w", "16a4w", "16a4w_block", "16a16w",):
+            fixed_point_type["io_type"] = torch.uint16
+            fixed_point_type["kv_type"] = torch.uint16
```

Comment:
For 4bit weights do you always use 16 bit activations and kv cache? You dont quantize kv cache to 8 bit in this case?